IdentityCore/src/cache/mac/MSIDMacTokenCache.m

```diff
@@ -22,39 +22,448 @@
 // THE SOFTWARE.
 
 #import "MSIDMacTokenCache.h"
+#import "MSIDToken.h"
+#import "MSIDTokenCacheKey.h"
+#import "MSIDTokenSerializer.h"
+#import "MSIDUserInformation.h"
+
+#define CURRENT_WRAPPER_CACHE_VERSION 1.0
+
+#define RETURN_ERROR_IF_CONDITION_FALSE(_cond, _code, _details) { \
+    if (!(_cond)) { \
+        NSError* _MSID_ERROR = MSIDCreateError(MSIDErrorDomain, _code, _details, nil, nil, nil, nil, nil); \
+        if (error) { *error = _MSID_ERROR; } \
+        return NO; \
+    } \
+}
+
+@interface MSIDMacTokenCache ()
+
+@property (nonatomic) NSMutableDictionary *cache;
+@property (nonatomic) dispatch_queue_t synchronizationQueue;
+
+@end
 
 @implementation MSIDMacTokenCache
 
-- (BOOL)setItemWithKey:(MSIDTokenCacheKey *)key
-                  item:(MSIDToken *)item
-            serializer:(id<MSIDTokenSerializer>)serializer
-               context:(id<MSIDRequestContext>)context
-                 error:(NSError **)error
+- (dispatch_queue_t)synchronizationQueue
 {
-    return NO;
+    if (!_synchronizationQueue) {
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.msidmactokencache-%@", [NSUUID UUID].UUIDString];
+        _synchronizationQueue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
+    }
+    
+    return _synchronizationQueue;
+}
+
++ (MSIDMacTokenCache *)defaultCache
+{
+    static dispatch_once_t once;
+    static MSIDMacTokenCache *cache = nil;
+    
+    dispatch_once(&once, ^{
+        cache = [MSIDMacTokenCache new];
+    });
+    
+    return cache;
+}
+
+- (nullable NSData *)serialize
+{
+    if (!self.cache)
+    {
+        return nil;
+    }
+    
+    __block NSDictionary *cacheCopy = nil;
+    dispatch_sync(self.synchronizationQueue, ^{
+        cacheCopy = [self.cache mutableCopy];
+    });
+    
+    NSData *result = nil;
+    
+    // Using the dictionary @{ key : value } syntax here causes _cache to leak. Yay legacy runtime!
+    NSDictionary *wrapper = [NSDictionary dictionaryWithObjectsAndKeys:cacheCopy, @"tokenCache",@CURRENT_WRAPPER_CACHE_VERSION, @"version", nil];
+    
+    @try
+    {
+        NSMutableData *data = [NSMutableData data];
+        
+        NSKeyedArchiver *archiver = [[NSKeyedArchiver alloc] initForWritingWithMutableData:data];
+        // Maintain backward compatibility with ADAL.
+        [archiver setClassName:@"ADTokenCacheKey" forClass:MSIDTokenCacheKey.class];
+        [archiver setClassName:@"ADTokenCacheStoreItem" forClass:MSIDToken.class];
+        [archiver setClassName:@"ADUserInformation" forClass:MSIDUserInformation.class];
+        [archiver encodeObject:wrapper forKey:NSKeyedArchiveRootObjectKey];
+        [archiver finishEncoding];
+        
+        result = data;
+    }
+    @catch (id exception)
```

Comment:
Just curious, why exception is of "id" type?

Reply:
Just copied it from ADTokenCache. Probably it is 'id' because we don't use it and don't care about it's type.

Reply:
I see, I'd prefer NSException here, but it's not that important.

Reply:
Actually, I recalled that in c++ you can throw any object, even number. I played with it in obj-c and it works as well. So I think it safer to have id there :) The following example will print 'id':

        @try {
            @throw @1;
        } @catch (NSException *e) {
            NSLog(@"NSException");
        } @catch (id e) {
            NSLog(@"id");
        }

Reply:
If I recall correctly, NSKeyedArchiver should only throw NSInvalidArchiveOperationException, so this won't be the case here?
Again, I'm fine with either ways, was just curious...